apisix/plugins/csrf.lua

```diff
@@ -0,0 +1,161 @@
+-- this work for additional information regarding copyright ownership.
+-- The ASF licenses this file to You under the Apache License, Version 2.0
+-- (the "License"); you may not use this file except in compliance with
+-- the License.  You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+local core = require("apisix.core")
+local resty_sha256 = require("resty.sha256")
+local str = require("resty.string")
+local ngx = ngx
+local ngx_encode_base64 = ngx.encode_base64
+local ngx_decode_base64 = ngx.decode_base64
+local ngx_time = ngx.time
+local cookie_time = ngx.cookie_time
+local math = math
+
+
+local schema = {
+    type = "object",
+    properties = {
+        key = {
+            description = "use to generate csrf token",
+            type = "string",
+        },
+        expires = {
+            description = "expires time(s) for csrf token",
+            type = "integer",
+            default = 7200
+        },
+        name = {
+            description = "the csrf token name",
+            type = "string",
+            default = "apisix-csrf-token"
+        }
+    },
+    required = {"key"}
+}
+
+
+local _M = {
+    version = 0.1,
+    priority = 2980,
+    name = "csrf",
+    schema = schema,
+}
+
+
+function _M.check_schema(conf)
+    return core.schema.check(schema, conf)
+end
+
+
+local function gen_sign(random, expires, key)
+    local sha256 = resty_sha256:new()
+
+    local sign = "{expires:" .. expires .. ",random:" .. random .. ",key:" .. key .. "}"
+
+    sha256:update(sign)
+    local digest = sha256:final()
+
+    return str.to_hex(digest)
+end
+
+
+local function gen_csrf_token(conf)
+    local random = math.random()
+    local sign = gen_sign(random, conf.expires, conf.key)
+
+    local token = {
+        random = random,
+        expires = conf.expires,
+        sign = sign,
+    }
+
+    local cookie = ngx_encode_base64(core.json.encode(token))
+    return cookie
+end
+
+
+local function check_csrf_token(conf, ctx, token)
+    local token_str = ngx_decode_base64(token)
+    if token_str == nil then
+        core.log.error("csrf token base64 decode error")
+        return false
+    end
+
+    local token_table, err = core.json.decode(token_str)
+    if err then
+        core.log.error("decode token error: ", err)
+        return false
+    end
+
+    local random = token_table["random"]
+    if not random then
+        core.log.error("no random in token")
+        return false
+    end
+
+    local expires = token_table["expires"]
+    if not expires then
+        core.log.error("no expires in token")
+        return false
+    end
```

Comment:
Just a proposal here, in the token object instead of total expiration seconds, can we use the expiration timestamp? So during the checking phase inside the `check_csrf_token` method, we could potentially check/prevent/[minimize the window] of replay attacks as an added benefit by figuring out if the generated cookie is long been expired or not. WDYT
Thank you. cc @spacewander

Reply:
Yes. Thanks for @bisakhmondal 's suggestion.

Reply:
Hey @Baoyuantop, thanks for making the changes. Are you going to address this in this PR or do you think it would be better to have it on a separate PR? Thank you.

Reply:
Hi @bisakhmondal, I would prefer a separate PR to solve this problem, this PR is already a bit big.

Reply:
Sure. Sounds good to me. Thanks

Reply:
Please open an issue for keeping track of the progress.

Reply:
Done @bisakhmondal. https://github.com/apache/apisix/issues/6141

Reply:
Thank you. Merging.